### PR TITLE
proposal: capability addition to ibc-go structure

### PR DESCRIPTION
**Pull request by tac0turtle**
## Description

This is a proposal on how to adopt capability. In the near future we will be able to remove the dependency of the sdk from this module making it standalone. The reason for the separate go.mod is that this module will most likely hit v1 and not iterate further with the plans of the sdk to introduce internal message routers. Tying it to an ibc-go release could be done as well, but didnt see the point since there wont be iteration on this work for a while. 

before going further wanted to double check if this is acceptable

closes: #3076

### Commit Message / Changelog Entry

```bash
feat: migrate capability module from cosmos-sdk to ibc-go
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/CONTRIBUTING.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/10-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/CONTRIBUTING.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

**Comment by tac0turtle** [comment]
any chance i can get feedback on this? would like to move forward while its still fresh in my mind

**Comment by crodriguezvega** [comment]
> any chance i can get feedback on this? would like to move forward while its still fresh in my mind

Thank you for helping us with this work, @tac0turtle. We'll try to review this or next week, if that's ok.

**Comment by damiannolan** [comment]
We can take over this PR and make any changes required to get it merged. 

- I think we need to update errors to use errorsmod instead of sdkerrors. 
- Copy over protos and regenerate (I think the current .pb.go files are copied directly)

I'd like to suggest we move `capability` from the modules directory to the root of the repo, and then tag as: `capability/v1.0.0`

**Comment by chatton** [comment]
We made the followup updates to this PR:

- Added proto files (and regenerated them).
- Moved capability module up one directory as @damiannolan suggested.
- Updated sdk to tip of rc2 release branch.
- Updated tendermint references to point at cometbft

We can't fully test/merge this until [this](https://github.com/strangelove-ventures/interchaintest/pull/359) PR is merged (which will be updating interchain test to cometbft) and then we can update ibc-go to use cometbft as well, which will then let us update references to this new module and run our E2E.

File: capability/keeper/keeper.go [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+func (k *Keeper) InitMemStore(ctx sdk.Context) {
+	memStore := ctx.KVStore(k.memKey)
+	memStoreType := memStore.GetStoreType()
+
+	if memStoreType != storetypes.StoreTypeMemory {
+		panic(fmt.Sprintf("invalid memory store type; got %s, expected: %s", memStoreType, storetypes.StoreTypeMemory))
+	}
+
+	// create context with no block gas meter to ensure we do not consume gas during local initialization logic.
+	noGasCtx := ctx.WithBlockGasMeter(storetypes.NewInfiniteGasMeter())
```
Comment (tac0turtle) [inline capability/keeper/keeper.go:122]: this is one pr we should adopt as well https://github.com/cosmos/cosmos-sdk/pull/15030

**Comment by chatton** [comment]
Myself and @damiannolan  did some more work to push this forward, we ran into a lot of issues which need further debugging but we were able to get everything building and tests passing using a local pin of the new go module.

1. Because we are using a local pin, we needed to modify the Dockerfile to `ADD` the modules directory before calling go mod download, this can be reverted when we are not on a local pin.
2. Capabilities errors were failing to register as the duplicate module/code pairs are already registered in the 0.47-rc3 tag of the sdk. As a workaround ,we have simply changed the codes.
3. We needed to pin the iavl module to `v0.20.0-alpha4`. (down from v0.21.0-alpha.1), without this pin, we get the following error

```
go build ./...
# github.com/cosmos/cosmos-sdk/store/iavl
vendor/github.com/cosmos/cosmos-sdk/store/iavl/tree.go:11:11: cannot use (*immutableTree)(nil) (value of type *immutableTree) as type Tree in variable declaration:
        *immutableTree does not implement Tree (wrong type for Iterator method)
                have Iterator(start []byte, end []byte, ascending bool) ("github.com/cosmos/cosmos-db".Iterator, error)
                want Iterator(start []byte, end []byte, ascending bool) ("github.com/cometbft/cometbft-db".Iterator, error)
vendor/github.com/cosmos/cosmos-sdk/store/iavl/tree.go:12:11: cannot use (*iavl.MutableTree)(nil) (value of type *"github.com/cosmos/iavl".MutableTree) as type Tree in variable declaration:
        *"github.com/cosmos/iavl".MutableTree does not implement Tree (wrong type for Iterator method)
                have Iterator(start []byte, end []byte, ascending bool) ("github.com/cosmos/cosmos-db".Iterator, error)
                want Iterator(start []byte, end []byte, ascending bool) ("github.com/cometbft/cometbft-db".Iterator, error)
vendor/github.com/cosmos/cosmos-sdk/store/iavl/store.go:55:43: cannot use db (variable of type "github.com/cometbft/cometbft-db".DB) as type "github.com/cosmos/cosmos-db".DB in argument to iavl.NewMutableTreeWithOpts:
        "github.com/cometbft/cometbft-db".DB does not implement "github.com/cosmos/cosmos-db".DB (wrong type for Iterator method)
                have Iterator(start []byte, end []byte) ("github.com/cometbft/cometbft-db".Iterator, error)
                want Iterator(start []byte, end []byte) ("github.com/cosmos/cosmos-db".Iterator, error)
vendor/github.com/cosmos/cosmos-sdk/store/iavl/store.go:90:11: cannot use tree (variable of type *"github.com/cosmos/iavl".MutableTree) as type Tree in struct literal:
        *"github.com/cosmos/iavl".MutableTree does not implement Tree (wrong type for Iterator method)
                have Iterator(start []byte, end []byte, ascending bool) ("github.com/cosmos/cosmos-db".Iterator, error)
                want Iterator(start []byte, end []byte, ascending bool) ("github.com/cometbft/cometbft-db".Iterator, error)
vendor/github.com/cosmos/cosmos-sdk/store/iavl/store.go:103:9: cannot use tree (variable of type *"github.com/cosmos/iavl".MutableTree) as type Tree in struct literal:
        *"github.com/cosmos/iavl".MutableTree does not implement Tree (wrong type for Iterator method)
                have Iterator(start []byte, end []byte, ascending bool) ("github.com/cosmos/cosmos-db".Iterator, error)
                want Iterator(start []byte, end []byte, ascending bool) ("github.com/cometbft/cometbft-db".Iterator, error)
vendor/github.com/cosmos/cosmos-sdk/store/iavl/store.go:123:9: cannot use &immutableTree{…} (value of type *immutableTree) as type Tree in struct literal:
        *immutableTree does not implement Tree (wrong type for Iterator method)
                have Iterator(start []byte, end []byte, ascending bool) ("github.com/cosmos/cosmos-db".Iterator, error)
                want Iterator(start []byte, end []byte, ascending bool) ("github.com/cometbft/cometbft-db".Iterator, error)
vendor/github.com/cosmos/cosmos-sdk/store/iavl/store.go:277:14: impossible type assertion: istore.tree.(*immutableTree)
        *immutableTree does not implement Tree (wrong type for Iterator method)
                have Iterator(start []byte, end []byte, ascending bool) ("github.com/cosmos/cosmos-db".Iterator, error)
                want Iterator(start []byte, end []byte, ascending bool) ("github.com/cometbft/cometbft-db".Iterator, error)
vendor/github.com/cosmos/cosmos-sdk/store/iavl/store.go:286:14: impossible type assertion: st.tree.(*iavl.MutableTree)
        *"github.com/cosmos/iavl".MutableTree does not implement Tree (wrong type for Iterator method)
                have Iterator(start []byte, end []byte, ascending bool) ("github.com/cosmos/cosmos-db".Iterator, error)
                want Iterator(start []byte, end []byte, ascending bool) ("github.com/cometbft/cometbft-db".Iterator, error)
``` 
4. We needed to pin `ledger-cosmos-go` to `v0.12.1` (the version sdk 47 uses). This was being indirectly pulled in at version `v0.13.0` 
The error we got before pinning was 

```
make build
go build -mod=readonly -tags "netgo ledger" -ldflags '-X github.com/cosmos/cosmos-sdk/version.Name=sim -X github.com/cosmos/cosmos-sdk/version.AppName=simd -X github.com/cosmos/cosmos-sdk/version.Version=02-client-refactor-beta1-171-gd9774d95d -X github.com/cosmos/cosmos-sdk/version.Commit=d9774d95dd61720cb9327c2d0630e879062cf822 -X "github.com/cosmos/cosmos-sdk/version.BuildTags=netgo,ledger" -w -s' -trimpath -o /Users/chatton/checkouts/ibc-go/build/ ./...
# github.com/cosmos/cosmos-sdk/crypto/ledger
../../go/pkg/mod/github.com/cosmos/cosmos-sdk@v0.47.0-rc3/crypto/ledger/ledger_real.go:20:10: cannot use device (variable of type *ledger_cosmos_go.LedgerCosmos) as type SECP256K1 in return statement:
        *ledger_cosmos_go.LedgerCosmos does not implement SECP256K1 (wrong type for SignSECP256K1 method)
                have SignSECP256K1(bip32Path []uint32, transaction []byte, p2 byte) ([]byte, error)
                want SignSECP256K1([]uint32, []byte) ([]byte, error)
make: *** [build] Error 2
```

5. Some of the unit tests are failing for unknown reasons and have been commented out. These need to be debugged and fixed.

**Review by chatton — COMMENTED** [review]
Changes seem fine to me, but I will leave final approval to someone else on the team who was not as involved in adding it!

Thanks @damiannolan / @tac0turtle

File: proto/capability/v1beta1/capability.proto [hunk 2]
```diff
@@ -0,0 +1,32 @@
+syntax = "proto3";
+
+package capability.v1beta1;
```
Comment (colin-axner) [inline proto/capability/v1beta1/capability.proto:3]: Should we be changing the package name? Should we use just `v1` if so?

File: proto/capability/v1beta1/capability.proto
Comment (colin-axner) [inline proto/capability/v1beta1/capability.proto:1]: for my understanding. The existing `module.proto` in the SDK is for app v2 dep inject stuff?

File: modules/capability/README.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+- [`x/capability`](#xcapability)
+  - [Overview](#overview)
+  - [Initialization](#initialization)
+  - [Contents](#contents)
+  - [Concepts](#concepts)
+    - [Capabilities](#capabilities)
+    - [Stores](#stores)
+  - [State](#state)
+    - [In persisted KV store](#in-persisted-kv-store)
+    - [In-memory KV store](#in-memory-kv-store)
```
Comment (colin-axner) [inline modules/capability/README.md:82]: looks like merge conflicts, concepts listed twice, should rename `x/capability` to `modules/capability`

File: modules/capability/go.mod [hunk 4]
```diff
@@ -0,0 +1,169 @@
+module github.com/cosmos/ibc-go/modules/capability
+
+go 1.19
+
+replace github.com/cosmos/ibc-go/v7 => ../../
```
Comment (colin-axner) [inline modules/capability/go.mod:5]: Why does capability module reference ibc-go? I guess for tests? Will this cause versioning issues?

File: modules/capability/capability_test.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+	ctx := suite.app.BaseApp.NewUncachedContext(false, tmproto.Header{})
+	newKeeper.Seal()
+	suite.Require().False(newKeeper.IsInitialized(ctx), "memstore initialized flag set before BeginBlock")
+
+	// Mock app beginblock and ensure that no gas has been consumed and memstore is initialized
+	ctx = suite.app.BaseApp.NewContext(false, tmproto.Header{}).WithBlockGasMeter(sdk.NewGasMeter(50))
+
+	prevBlockGas := ctx.BlockGasMeter().GasConsumed()
+	prevGas := ctx.BlockGasMeter().GasConsumed()
+
```
Comment (colin-axner) [inline modules/capability/capability_test.go:68]: merge conflict?

**Review by colin-axner — COMMENTED** [review]
Thank you @damiannolan and @chatton!!! Excellent work! :100: 

I have reviewed the diffs against SDK v0.47 release and I think I found a few merge conflicts that got duplicated. Otherwise everything looks great to me. I think it'd be nice if the go.mod didn't have to pin core IBC just for testing (will we remove this for releases?). I think this could be addressed by making simapp its own go.mod

File: modules/capability/capability_test.go (hunk 5 above)
Comment (chatton) [reply modules/capability/capability_test.go:68]: this was done by hand to fix the tests to make them use the old wiring (non-appv2)

File: modules/capability/capability_test.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+	restartedModule := capability.NewAppModule(suite.cdc, *newKeeper, false)
```
[the remark this answers is not in the document]
Comment (chatton) [reply modules/capability/capability_test.go:69]: the 47 code didn't have this fix, this was pulled in based on [this](https://github.com/cosmos/ibc-go/pull/3097#discussion_r1121876811).

[the remark this answers is not in the document]
Comment (chatton) [reply modules/capability/capability_test.go:69]: although it does still pass with the false set back to true, so we can make that change!

File: modules/capability/capability_test.go (hunk 5 above)
Comment (colin-axner) [reply modules/capability/capability_test.go:70]: I'm a bit confused? `prevBlockGas` and `prevGas` look like duplicates to me?

Comment (colin-axner) [reply modules/capability/capability_test.go:70]:
```
suite.Require().Equal(prevBlockGas, blockGasUsed, "ensure beginblocker consumed no block gas during execution")
suite.Require().Equal(prevGas, gasUsed, "ensure beginblocker consumed no gas during execution")
```

Comment (colin-axner) [reply modules/capability/capability_test.go:70]: same with `blockGasUsed` and `gasUsed`:
```go
gasUsed := ctx.GasMeter().GasConsumed()

	suite.Require().True(newKeeper.IsInitialized(ctx), "memstore initialized flag not set")
blockGasUsed := ctx.BlockGasMeter().GasConsumed()
```

File: modules/capability/capability_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (colin-axner) [reply modules/capability/capability_test.go]: ah I see now! The naming between `BlockGasMeter` and `GasMeter` is incredibly small! Really wish it was `BlockGasMeter` and `TxGasMeter`

File: modules/capability/keeper/keeper.go [hunk 7]
```diff
@@ -0,0 +1,494 @@
...
+// can't initialize it in a constructor.
+func (k *Keeper) InitMemStore(ctx sdk.Context) {
+	memStore := ctx.KVStore(k.memKey)
+	memStoreType := memStore.GetStoreType()
+
+	if memStoreType != storetypes.StoreTypeMemory {
+		panic(fmt.Sprintf("invalid memory store type; got %s, expected: %s", memStoreType, storetypes.StoreTypeMemory))
+	}
+
+	// create context with no block gas meter to ensure we do not consume gas during local initialization logic.
```
Comment (colin-axner) [inline modules/capability/keeper/keeper.go:121]:
```suggestion
	// create context with no block gas meter and no transaction gas meter to ensure we do not consume gas 
	// during local initialization logic.
```

**Review by colin-axner — APPROVED** [review]
Nice job y'all! Everything looks good to go to me!

**Pull request merged**